src/sagemaker/model.py

```diff
@@ -236,10 +244,32 @@ def _compilation_job_config(
         }
         role = self.sagemaker_session.expand_role(role)
         output_model_config = {
-            "TargetDevice": target_instance_type,
             "S3OutputLocation": output_path,
         }
 
+        if target_instance_type is not None:
+            output_model_config["TargetDevice"] = target_instance_type
+        else:
+            if target_platform_os is None and target_platform_arch is None:
+                raise ValueError(
+                    "target_instance_type or (target_platform_os and target_platform_arch) "
+                    "should be provided"
+                )
+            target_platform = {
+                "Os": target_platform_os,
+                "Arch": target_platform_arch,
+            }
+            if target_platform_accelerator is not None:
+                target_platform["Accelerator"] = target_platform_accelerator
+            output_model_config["TargetPlatform"] = target_platform
+
+        if compiler_options is not None:
+            output_model_config["CompilerOptions"] = (
+                compiler_options
+                if not isinstance(compiler_options, dict)
+                else json.dumps(compiler_options)
```

Comment:
let's flip this around to put the positive condition first, i.e. `json.dumps(compiler_options) if isinstance(compiler_options, dict) else compiler_options`

Reply:
fixed here and for `DataInputConfig` as well.